File: server/src/openproject/mappers.ts
import { idFromHref } from './client.js';

/* ─────────────────────────────────────────────────────────────
 * Domain DTOs
 * ───────────────────────────────────────────────────────────── */

export interface Task {
  id: string;
  subject: string;
  start: string | null;
  due: string | null;
  progress: number;
  status: string;
  typeName: string;
  assigneeId: string | null;
  assigneeType: 'user' | 'group' | null;
  parentId: string | null;
  projectId: string | null;
  dependencies: string[];
}

export interface User {
  id: string;
  name: string;
  email: string | null;
  avatarUrl: string | null;
  groupIds: string[];
}

export interface Group {
  id: string;
  name: string;
  memberIds: string[];
}

/* ─────────────────────────────────────────────────────────────
 * Mapping functions  – raw HAL → domain DTO
 * ───────────────────────────────────────────────────────────── */

export function mapWorkPackage(wp: Record<string, any>): Task {
  const assigneeHref: string | null = wp._links?.assignee?.href ?? null;

  let assigneeType: Task['assigneeType'] = null;
  if (assigneeHref?.includes('/groups/')) {
    assigneeType = 'group';
  } else if (assigneeHref?.includes('/users/')) {
    assigneeType = 'user';
  }

  return {
    id: String(wp.id),
    subject: wp.subject,
    start: wp.startDate ?? null,
    due: wp.dueDate ?? null,
    progress: wp.percentageDone ?? 0,
    status: wp._links?.status?.title ?? '',
    typeName: wp._links?.type?.title ?? '',
    assigneeId: idFromHref(assigneeHref),
    assigneeType,
    parentId: idFromHref(wp._links?.parent?.href ?? null),
    projectId: idFromHref(wp._links?.project?.href ?? null),
    dependencies: [],
  };
}

export function mapUser(u: Record<string, any>): User {
  return {
    id: String(u.id),
    name: u.name,
    email: u.email ?? null,
    avatarUrl: u._links?.avatar?.href ?? null,
    groupIds: [],
  };
}

export function mapGroup(g: Record<string, any>): Group {
  let memberIds: string[] = [];

  if (g._embedded?.members) {
    memberIds = g._embedded.members.map(
      (m: Record<string, any>) => idFromHref(m._links?.self?.href),
    ).filter(Boolean) as string[];
  } else if (g._links?.members) {
    const members: Array<{ href: string }> = g._links.members;
    memberIds = members.map(m => idFromHref(m.href)).filter(Boolean) as string[];
  }

  return {
    id: String(g.id),
    name: g.name,
    memberIds,
  };
}

export interface Project {
  id: string;
  name: string;
}

export function mapProject(p: Record<string, any>): Project {
  return {
    id: String(p.id),
    name: p.name,
  };
}

File: server_py/app/openproject/mappers.py
from typing import Any

from app.openproject.client import id_from_href
from app.openproject.schemas import Group, Project, Task, User


def map_work_package(wp: dict[str, Any]) -> Task:
    links = wp.get("_links") or {}
    assignee_href = (links.get("assignee") or {}).get("href")

    assignee_type = None
    if assignee_href and "/groups/" in assignee_href:
        assignee_type = "group"
    elif assignee_href and "/users/" in assignee_href:
        assignee_type = "user"

    return Task(
        id=str(wp.get("id")),
        subject=wp.get("subject"),
        start=wp.get("startDate"),
        due=wp.get("dueDate"),
        progress=wp.get("percentageDone") or 0,
        status=(links.get("status") or {}).get("title") or "",
        typeName=(links.get("type") or {}).get("title") or "",
        assigneeId=id_from_href(assignee_href),
        assigneeType=assignee_type,
        parentId=id_from_href((links.get("parent") or {}).get("href")),
        projectId=id_from_href((links.get("project") or {}).get("href")),
        dependencies=[],
    )


def map_user(user: dict[str, Any]) -> User:
    links = user.get("_links") or {}
    return User(
        id=str(user.get("id")),
        name=user.get("name"),
        email=user.get("email"),
        avatarUrl=(links.get("avatar") or {}).get("href"),
        groupIds=[],
    )


def map_group(group: dict[str, Any]) -> Group:
    member_ids: list[str] = []
    embedded_members = (group.get("_embedded") or {}).get("members")

    if embedded_members:
        member_ids = [
            member_id
            for member in embedded_members
            if (member_id := id_from_href(((member.get("_links") or {}).get("self") or {}).get("href")))
        ]
    elif (group.get("_links") or {}).get("members"):
        member_ids = [
            member_id
            for member in (group.get("_links") or {}).get("members", [])
            if (member_id := id_from_href(member.get("href")))
        ]

    return Group(
        id=str(group.get("id")),
        name=group.get("name"),
        memberIds=member_ids,
    )


def map_project(project: dict[str, Any]) -> Project:
    return Project(
        id=str(project.get("id")),
        name=project.get("name"),
    )
